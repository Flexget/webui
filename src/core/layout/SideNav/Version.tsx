import React, { FC } from 'react';
import { gt } from 'semver';
import { css } from '@emotion/core';
import { IconButton, Theme } from '@material-ui/core';
import HelpOutlineIcon from '@material-ui/icons/HelpOutline';
import Typography from '@material-ui/core/Typography';
import { useVersion } from './hooks';

interface Props {
  className?: string;
}

const wrapper = (theme: Theme) => css`
  color: ${theme.palette.primary.main};
  opacity: 1;
  padding: ${theme.typography.pxToRem(theme.spacing(2))};
`;

const Version: FC<Props> = ({ className }) => {
  const { loading, version } = useVersion();

  if (loading || !version) {
    // showProgress
    return null;
  }

  const { flexgetVersion, apiVersion, latestVersion } = version;

  const normalizedVersion = flexgetVersion.replace('.dev', '-dev.0');

  return (
    <div css={wrapper} className={className}>
      <Typography>Version Info</Typography>
      <Typography>
        {`Flexget: ${flexgetVersion} `}
        {gt(latestVersion, normalizedVersion) && (
          <IconButton
            href="https://flexget.com/ChangeLog"
            color="inherit"
            size="small"
            aria-label="flexget update available"
          >
            <HelpOutlineIcon />
          </IconButton>
        )}
      </Typography>
      <Typography>{`API: ${apiVersion}`}</Typography>
    </div>
  );
};

export default Version;
