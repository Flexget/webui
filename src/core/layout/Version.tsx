import React, { useEffect, useState } from 'react';
import semver from 'semver-compare';
import styled from '@emotion/styled';
import theme from 'theme';
import { IconButton } from '@material-ui/core';
import HelpOutlineIcon from '@material-ui/icons/HelpOutline';
import Typography from '@material-ui/core/Typography';
import { useFlexgetAPI } from 'core/api';
import { useGlobalStatus } from 'core/status/hooks';

interface Props {
  className?: string;
}

interface VersionResponse {
  apiVersion: string;
  flexgetVersion: string;
  latestVersion: string;
}

const Wrapper = styled.div`
  color: ${theme.palette.primary[800]};
`;

const Line = styled(Typography)`
  margin: 0;
`;

const useVersion = () => {
  const [version, setVersion] = useState<VersionResponse | undefined>();
  const [{ loading, error }, getVersion] = useFlexgetAPI<VersionResponse>('/server/version');
  useGlobalStatus(loading, error);

  useEffect(() => {
    const fn = async () => {
      const resp = await getVersion();
      if (resp.ok) {
        setVersion(resp.data);
      }
    };
    fn();
  }, [getVersion]);

  return { loading, version };
};

const Version: React.FC<Props> = ({ className }) => {
  const { loading, version } = useVersion();

  if (loading || !version) {
    // showProgress
    return null;
  }

  const { flexgetVersion, apiVersion, latestVersion } = version;

  return (
    <Wrapper className={className}>
      <Line>Version Info</Line>
      <Line>
        {`Flexget: ${flexgetVersion} `}
        {semver(latestVersion, flexgetVersion) !== 1 && (
          <IconButton href="https://flexget.com/ChangeLog" color="inherit" size="small">
            <HelpOutlineIcon />
          </IconButton>
        )}
      </Line>
      <Line>{`API: ${apiVersion}`}</Line>
    </Wrapper>
  );
};

export default Version;
