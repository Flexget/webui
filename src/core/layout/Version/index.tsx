import React from 'react';
import semver from 'semver-compare';
import styled from '@emotion/styled';
import { useFlexgetAPI } from 'utils/hooks/api';
import theme from 'theme';
import { IconButton } from '@material-ui/core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface Props {
  className?: string;
}

interface VersionResponse {
  apiVersion: string;
  flexgetVersion: string;
  latestVersion: string;
}

const Wrapper = styled.div`
  color: ${theme.palette.primary[800]};
`;

const Line = styled.p`
  margin: 0;
`;

const Version: React.FC<Props> = ({ className }) => {
  const { loading, error, data } = useFlexgetAPI<VersionResponse>('/server/version', []);

  if (loading || !data) {
    // showProgress
    return null;
  }

  if (error) {
    return null;
  }

  const { flexgetVersion, apiVersion, latestVersion } = data;

  return (
    <Wrapper className={className}>
      <Line>Version Info</Line>
      <Line>
        {`Flexget: ${flexgetVersion} `}
        {semver(latestVersion, flexgetVersion) === 1 && (
          <IconButton href="https://flexget.com/ChangeLog">
            <FontAwesomeIcon icon={['far', 'question-circle']} fixedWidth />
          </IconButton>
        )}
      </Line>
      <Line>{`API: ${apiVersion}`}</Line>
    </Wrapper>
  );
};

export default Version;
