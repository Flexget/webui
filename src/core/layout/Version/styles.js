import styled from '@emotion/styled';
import theme from 'theme';

export const Wrapper = styled.div`
  color: ${theme.palette.primary[800]};
`;

export const Line = styled.p`
  margin: 0;
`;
