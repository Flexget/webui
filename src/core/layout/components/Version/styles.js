import styled from 'react-emotion';
import theme from 'theme';

export const Wrapper = styled.div`
  color: ${theme.palette.error[500]};
`;

export const Line = styled.p`
  margin: 0;
`;
