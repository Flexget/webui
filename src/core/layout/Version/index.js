import { connect } from 'react-redux';
import actions from 'core/version/state/actions';
import Version from './Version';

export function mapStateToProps({ version }) {
  return {
    version,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    getVersion: () => dispatch(actions.getVersion.request()),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Version);
export { Version };
