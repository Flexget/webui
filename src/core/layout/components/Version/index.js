import { connect } from 'react-redux';
import { GET_VERSION } from 'core/version/data/actions';
import { request } from 'utils/actions';
import Version from './Version';

export function mapStateToProps({ version }) {
  return {
    version,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    getVersion: () => dispatch(request(GET_VERSION)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Version);
export { Version };
