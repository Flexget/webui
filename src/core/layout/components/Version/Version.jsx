import React, { Component } from 'react';
import PropTypes from 'prop-types';
import semver from 'semver-compare';
import IconButton from 'material-ui/IconButton';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import { Wrapper, Line } from './styles';

class Version extends Component {
  static propTypes = {
    version: PropTypes.shape({
      api: PropTypes.string,
      flexget: PropTypes.string,
      latest: PropTypes.string,
    }).isRequired,
    getVersion: PropTypes.func.isRequired,
    className: PropTypes.string,
  };

  static defaultProps = {
    className: '',
  };

  componentDidMount() {
    const { version, getVersion } = this.props;
    if (process.env.NODE_ENV === 'production' || !version.api) {
      getVersion();
    }
  }

  render() {
    const { version: { api, flexget, latest }, className } = this.props;
    return (
      <Wrapper className={className}>
        <Line>Version Info</Line>
        <Line>
Flexget:
          { flexget }
          {' '}
          {
          latest && semver(latest, flexget) === 1 && (
            <IconButton href="https://flexget.com/ChangeLog">
              <FontAwesomeIcon icon={['far', 'question-circle']} fixedWidth />
            </IconButton>
          ) }
        </Line>
        <Line>
API:
          { api }
        </Line>
      </Wrapper>
    );
  }
}

export default Version;
