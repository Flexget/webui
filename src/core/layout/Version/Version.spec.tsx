import React from 'react';
import { act, create, ReactTestRenderer } from 'react-test-renderer';
import Version from 'core/layout/Version';
import { themed } from 'utils/tests';
import { FetchMock } from 'jest-fetch-mock';

const fetchMock = fetch as FetchMock;

describe('core/layout/Version', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });
  it('renders correctly with latest version', async () => {
    fetchMock.mockResponseOnce(
      JSON.stringify({
        apiVersion: '1.1.2',
        flexgetVersion: '2.10.60',
        latestVersion: '2.10.60',
      }),
    );
    let tree: ReactTestRenderer | undefined;
    await act(async () => {
      tree = create(themed(<Version />));
    });
    expect(tree && tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly without latest version', async () => {
    fetchMock.mockResponseOnce(
      JSON.stringify({
        apiVersion: '1.1.2',
        flexgetVersion: '2.10.11',
        latestVersion: '2.10.60',
      }),
    );
    let tree: ReactTestRenderer | undefined;
    await act(async () => {
      tree = create(themed(<Version />));
    });
    expect(tree && tree.toJSON()).toMatchSnapshot();
  });
});
