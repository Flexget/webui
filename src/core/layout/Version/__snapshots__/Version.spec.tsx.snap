// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/layout/Version renders correctly with latest version 1`] = `
.emotion-3 {
  color: #ef6c00;
}

.emotion-0 {
  margin: 0;
}

<div
  className="emotion-3"
>
  <p
    className="emotion-0"
  >
    Version Info
  </p>
  <p
    className="emotion-0"
  >
    Flexget: 2.10.60 
  </p>
  <p
    className="emotion-0"
  >
    API: 1.1.2
  </p>
</div>
`;

exports[`core/layout/Version renders correctly without latest version 1`] = `
.emotion-3 {
  color: #ef6c00;
}

.emotion-0 {
  margin: 0;
}

<div
  className="emotion-3"
>
  <p
    className="emotion-0"
  >
    Version Info
  </p>
  <p
    className="emotion-0"
  >
    Flexget: 2.10.11 
    <a
      aria-disabled={false}
      className="MuiButtonBase-root MuiIconButton-root"
      href="https://flexget.com/ChangeLog"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      <span
        className="MuiIconButton-label"
      />
      <span
        className="MuiTouchRipple-root"
      />
    </a>
  </p>
  <p
    className="emotion-0"
  >
    API: 1.1.2
  </p>
</div>
`;
